### PR TITLE
Update advanced user/developer guides

### LINKED ISSUE #4840: Streamline steps for building and maintaining GMT
**Description of the desired feature**

The steps to build and run tests in GMT to produce new postscript (*.ps) files is not very intuitive for first timers. In particular, it's a little confusing having to jump between the [BUILDING.md](https://github.com/GenericMappingTools/gmt/blob/master/BUILDING.md) and [MAINTENANCE.md](https://github.com/GenericMappingTools/gmt/blob/master/MAINTENANCE.md) documents.

> Yes, I think from this it's clear from this PR that some revisions are needed to the building and maintaining (and possible contributing) documentation. When I started, I also found that I needed to bounce back and forth between building and maintaining documents a bunch. Perhaps some redundancy is needed for improved clarity. Would you want to open an issue for this?

_Originally posted by @meghanrjones in https://github.com/GenericMappingTools/gmt/issues/4838#issuecomment-783445471_

See also https://github.com/GenericMappingTools/gmt/pull/4838#issuecomment-782954781 and https://github.com/GenericMappingTools/gmt/pull/4838#issuecomment-782964119

Suggested changes:

- Consolidate BUILDING.md and MAINTENANCE.md together into one document, or maybe add some redundancy to both documents so that it will be clearer to first timers.
- Mention the convenience scripts at `admin/bashrc_for_gmt`(https://github.com/GenericMappingTools/gmt/blob/master/admin/bashrc_for_gmt) (added in #4529) in MAINTENANCE.md, and briefly mention how to use them

**Are you willing to help implement and maintain this feature?** Happy to review the Pull Request!

### PR DESCRIPTION
**Pull request by meghanrjones**
**Description of proposed changes**

This PR adds some more information to the contributing, maintenance, and building guides to help new developers.

Fixes #4840

**Reminders**

- [ ] Make sure that your code follows our style. Use the other functions/files as a basis.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Describe changes to function behavior and arguments in a comment below the function declaration.
- [ ] If adding new functionality, add a detailed description to the documentation and/or an example.

### CODE REVIEW

File: BUILDING.md [hunk 1]
```diff
@@ -19,12 +19,20 @@ GMT is dependent on some software and libraries to run.
 Please refer to the [GMT wiki page](https://github.com/GenericMappingTools/gmt/wiki)
 for instructions to install these dependencies on various operation systems.
 
+### Required dependencies
+
 To build GMT, you have to install:
```
Comment (weiji14) [inline BUILDING.md:24]: Is it worth listing [Ninja](https://ninja-build.org/) as an optional dependency? I see it mentioned sometimes as a faster way to build GMT.

File: MAINTENANCE.md [hunk 2]
```diff
@@ -23,50 +31,92 @@ cmake --build . --target docs_man         # UNIX manual pages
...
+
+        ctest --rerun-failed --verbose
+
+### Updating reference plots for tests
+
+Pull requests should avoid needing to change PostScript files in the [test/](/test) and [doc/scripts/](/doc/scripts)
+directories. However, if this is unavoidable, new PostScript reference files can be generated by running
+`ctest -R <test-script-name>` in the build directory after following the [building guide](BUILDING.md) and the
+[Configuring CMake for testing GMT](MAINTENANCE.md#configuring-cmake-for-testing-gmt) instructions. The new PostScript
+file can be copied from the appropriate subdirectory within `build/test/` or `build/doc/scipts/`.
```
Comment (weiji14) [inline MAINTENANCE.md:114]:
```suggestion
file can then be copied from the appropriate subdirectory within `build/test/` or `build/doc/scripts/`
to [test/](/test) or [doc/scripts/](/doc/scripts) respectively.
```

**Review by weiji14 — COMMENTED** [review]
Nice one! Just a few comments for now :smile:

File: BUILDING.md [hunk 3]
```diff
@@ -136,6 +158,8 @@ For advanced users, you can append the option `-G Ninja` to use the
 build tool [Ninja](https://ninja-build.org/), which is a small build system
 with a focus on speed.
 
+> Note for developers: Refer to the file `admin/bashrc_for_gmt` for useful aliases for configuring and building GMT.
```
[the remark this answers is not in the document]
Comment (meghanrjones) [reply BUILDING.md:162]: Let me know if you think the new section in MAINTENANCE.md is worthwhile. Mostly everything is also included as comments in `bashrc_for_gmt`, but perhaps this is more accessible.

File: BUILDING.md (hunk 1 above)
Comment (meghanrjones) [reply BUILDING.md:24]: Yes, now included.

File: MAINTENANCE.md [hunk 4]
```diff
@@ -163,3 +164,46 @@ the build directory you have to run CMake again manually.
...
+
+## Using build and test aliases
+
+The file [admin/bashrc_for_gmt](/admin/bashrc_for_gmt) contains useful aliases for building and testing GMT that some
+developers chose to use. New pull requests with other aliases that you find helpful are welcome. This file is version controlled, so you should copy the file to a different location in order to edit and use it. For example, use these
+commands to copy it to your home directory:
+
+```
+cd <path-to-gmt>
+cp admin/bashrc_for_gmt ~/bashrc_for_gmt
```
Comment (weiji14) [inline MAINTENANCE.md:176]: Not sure about bashrc best practices, but would it be better to make it a hidden file (with a dot `.`)?

```suggestion
cp admin/bashrc_for_gmt ~/.bashrc_for_gmt
```

File: MAINTENANCE.md [hunk 5]
```diff
@@ -163,3 +164,46 @@ the build directory you have to run CMake again manually.
...
+- Optionally, change `ncores=4` to the number of cores to use for building and running tests.
+- Change `MATLAB=/Applications/MATLAB_R2019a.app` to the path for your version of the MATLAB app.
+- Set `REPO_DIR` to the path that contains the GMT repository.
+- Set `DATA_DIR` to the path that contains the folders `dcw-gmt-1.1.4/` and `gshhg-gmt-2.3.7/` for the dcw and gshhg
+  dataset respectively. If these folders are not located in the same path, you can instead delete L25
+  (`DATA_DIR=<path to directory containing GSHHG and DCW>`) and set the individual paths to the GSHHG and DCW source
+  in L30 (`export GMT_GSHHG_SOURCE=${DATA_DIR}/gshhg-gmt-2.3.7`) and
+  L31(`export GMT_DCW_SOURCE=${DATA_DIR}/dcw-gmt-1.1.4`).
+- Edit the file `~/.bashrc` to include the line `source <path>/bashrc_for_gmt`. If you set up `bashrc_for_gmt` in your
+  home directory, this line should be `source ~/bashrc_for_gmt`.
```
Comment (weiji14) [inline MAINTENANCE.md:193]: Wondering if this bullet point should probably be on the top of the list. Or if it's better to have people configure things properly first.

File: MAINTENANCE.md [hunk 6]
```diff
@@ -96,13 +148,62 @@ cpack -G TGZ|TBZ2|Bundle|ZIP|NSIS
...
+## Debugging GMT
+
+Guides for debugging GMT are provided in the
+[developer resources section](https://docs.generic-mapping-tools.org/dev/debug.html) of the GMT documentation.
+
+## Using build and test aliases
+
+The file [admin/bashrc_for_gmt](/admin/bashrc_for_gmt) contains useful aliases for building and testing GMT that some
+developers chose to use. New pull requests with other aliases that you find helpful are welcome. This file is version controlled, so you should copy the file to a different location in order to edit and use it. For example, use these
+commands to copy it to your home directory:
```
Comment (weiji14) [inline MAINTENANCE.md:172]: Small nitpick to reduce line length.

```suggestion
developers chose to use. New pull requests with other aliases that you find helpful are welcome.
This file is version controlled, so you should copy the file to a different location in order to edit and use it.
For example, use these commands to copy it to your home directory:
```

File: MAINTENANCE.md [hunk 7]
```diff
@@ -96,13 +148,62 @@ cpack -G TGZ|TBZ2|Bundle|ZIP|NSIS
...
+- Set `REPO_DIR` to the path that contains the GMT repository.
+- Set `DATA_DIR` to the path that contains the folders `dcw-gmt-1.1.4/` and `gshhg-gmt-2.3.7/` for the dcw and gshhg
+  datasets respectively. If these folders are not located in the same path, you can instead delete L25
+  (`DATA_DIR=<path to directory containing GSHHG and DCW>`) and set the individual paths to the GSHHG and DCW source
+  in L30 (`export GMT_GSHHG_SOURCE=${DATA_DIR}/gshhg-gmt-2.3.7`) and
+  L31 (`export GMT_DCW_SOURCE=${DATA_DIR}/dcw-gmt-1.1.4`).
+- Edit the file `~/.bashrc` to include the line `source <path>/bashrc_for_gmt`. If you set up `bashrc_for_gmt` in your
+  home directory, this line should be `source ~/bashrc_for_gmt`.
+
+Here are some of the shortcuts included in `bashrc_for_gmt`:
```
Comment (weiji14) [inline MAINTENANCE.md:195]:
```suggestion
Here are some of the shortcuts included in [`bashrc_for_gmt`](/admin/bashrc_for_gmt):
```

File: MAINTENANCE.md [hunk 8]
```diff
@@ -96,13 +148,62 @@ cpack -G TGZ|TBZ2|Bundle|ZIP|NSIS
...
+The file [admin/bashrc_for_gmt](/admin/bashrc_for_gmt) contains useful aliases for building and testing GMT that some
+developers chose to use. New pull requests with other aliases that you find helpful are welcome. This file is version controlled, so you should copy the file to a different location in order to edit and use it. For example, use these
+commands to copy it to your home directory:
+
+```
+cd <path-to-gmt>
+cp admin/bashrc_for_gmt ~/bashrc_for_gmt
+```
+
+Here are the steps for setting up `bashrc_for_gmt` after copying it to a new location:
```
Comment (weiji14) [inline MAINTENANCE.md:179]:
```suggestion
Here are the steps for setting up [`bashrc_for_gmt`](/admin/bashrc_for_gmt) after copying it to a new location:
```

File: MAINTENANCE.md [hunk 9]
```diff
@@ -96,13 +148,62 @@ cpack -G TGZ|TBZ2|Bundle|ZIP|NSIS
...
+
+- If you do not have [ninja](https://ninja-build.org/) installed, you will need to change `builder=ninja` to
+  `builder=make` and `Bname="Ninja"` to `Bname="Unix Makefiles"`. It is recommended that you use ninja.
```
Comment (weiji14) [inline MAINTENANCE.md:182]:
```suggestion
  `builder=make` and `Bname="Ninja"` to `Bname="Unix Makefiles"`. Ninja is recommended for speeding up build times.
```

File: MAINTENANCE.md [hunk 10]
```diff
@@ -96,13 +148,62 @@ cpack -G TGZ|TBZ2|Bundle|ZIP|NSIS
...
+- You may need to update `pngview=open` and `pngview=open` depending on your preferred program for viewing files.
+- Optionally, change `ncores=4` to the number of cores to use for building and running tests.
+- Change `MATLAB=/Applications/MATLAB_R2019a.app` to the path for your version of the MATLAB app.
+- Set `REPO_DIR` to the path that contains the GMT repository.
```
Comment (weiji14) [inline MAINTENANCE.md:186]:
```suggestion
- Set `REPO_DIR` to the path that contains the local `git clone` copy of the GMT repository.
```

File: MAINTENANCE.md [hunk 11]
```diff
@@ -96,13 +148,62 @@ cpack -G TGZ|TBZ2|Bundle|ZIP|NSIS
...
+- Set `DATA_DIR` to the path that contains the folders `dcw-gmt-1.1.4/` and `gshhg-gmt-2.3.7/` for the dcw and gshhg
+  datasets respectively. If these folders are not located in the same path, you can instead delete L25
+  (`DATA_DIR=<path to directory containing GSHHG and DCW>`) and set the individual paths to the GSHHG and DCW source
+  in L30 (`export GMT_GSHHG_SOURCE=${DATA_DIR}/gshhg-gmt-2.3.7`) and
+  L31 (`export GMT_DCW_SOURCE=${DATA_DIR}/dcw-gmt-1.1.4`).
```
Comment (weiji14) [inline MAINTENANCE.md:191]: I worry that the line numbers may change in the future, so maybe remove the L25/L30/L31? What do you think, keep or remove?

**Review by weiji14 — COMMENTED** [review]
> Let me know if you think the new section in MAINTENANCE.md is worthwhile. Mostly everything is also included as comments in bashrc_for_gmt, but perhaps this is more accessible.

Looks great @meghanrjones! I think that listing down a few shortcut examples (e.g. `gmt6`, `gtop`, etc) as you've done is helpful since the bashrc_for_gmt file is quite long (>100 lines). Also had a look through the changes and I've got some questions/suggestions.

File: MAINTENANCE.md (hunk 5 above)
Comment (meghanrjones) [reply MAINTENANCE.md]: I think setting up first is better.

File: MAINTENANCE.md (hunk 11 above)
Comment (meghanrjones) [reply MAINTENANCE.md]: Yes, I removed the line numbers.

**Review by PaulWessel — APPROVED** [review]

**Pull request merged**